File: src/main/OSFI/Python/ParamType.py
#
# openSF Integration Libraries (OSFI)
# Deimos Space, S.L.U.
# 
# This file is part of OSFI. OSFI is free software; you can redistribute it
# and/or modify it under the terms of the 'ESA Software Community Licence Permissive' as
# published by the European Space Agency; either version 2.4 of the License,
# or (at your option) any later version. You should have received a
# copy of the 'ESA Software Community Licence Permissive - v2.4' along with this program
# or one can be found at <http://eop-cfi.esa.int/index.php/docs-and-mission-data/licensing-documents>.
#

import os
import re
import pathlib

try:
    from . import Logger
    from .TimeValue import TimeValue
    from .OsfiEnv import get_env
except ImportError: # XXX Deprecation of non-package mode
    import Logger
    from TimeValue import TimeValue
    from OsfiEnv import get_env


class ParamType(object):
    """
     Internal definition of parameter types
     """
    __values = {} # Keep a central DB of all param types
    
    def __new__(cls, name, parseFun, is_str):
        if name in ParamType.__values:
            raise ValueError("ParamType with that name already exists")
        instance = super(ParamType, cls).__new__(cls)
        ParamType.__values[name] = instance
        return instance

    def __init__(self, name, parseFun, is_str):
        self.__name = name
        self.__parseSingleVal = parseFun
        self.__isStringBased  = is_str
    
    def __str__(self):
        return self.__name
    
    def parseSingleValue(self, x):
        return self.__parseSingleVal(x)
    
    def parseMultipleValues(self, value, expectedLen=None):
        """
        Parse a sequence of values in a string into a vector. If expectedLen is given,
        also assert that the number of parsed values equals that argument.
        """
        vector = []
        tokens = self.splitValues(value, expectedLen is not None and expectedLen == 0)
        if expectedLen is not None and len(tokens) != expectedLen:
            raise ValueError("Number of elements is different than expected: {0} vs {1}"
                    .format(len(tokens), expectedLen))
        for iT, token in enumerate(tokens):
            try:
                vector.append(self.__parseSingleVal(token))
            except Exception as e:
                raise ValueError("Invalid value at element {0} for type {1}: '{2}'"
                        .format(iT, self.__name, str(e)))
        return vector
    
    # TODO: this was moved from Parameter.splitValues, it needs to be cleaned up
    def splitValues(self, string, expected_empty=False):
        """
        Split values into the tokens to be parsed
        @param string unparsed str or None (only for empty dimensions in ARRAY parameters)
        @param expected_empty True if the data corresponds to an empty dimension in an ARRAY parameter
        @return List[str] with the unparsed but split values
        """

        if (string is None or not string.strip()) and expected_empty:
            return []
        elif self.__isStringBased:
            # String should be in the format ""'a' 'b c' 'd'". The old algorithm (which is kept with
            # the substitution of str methods for re methods) is to split by "' '" and then to
            # remove the initial and final quotes.
            splitValues = re.split(r"'\s+'", string) # Note that "''" is not a valid delimiter
            firstQuote = re.match(r"\s*'", splitValues[0]) # "Match" is like an automatic ^ in the RE...
            if firstQuote:
                splitValues[0] = splitValues[0][firstQuote.end():]
            else: # A single unquoted value, remove leading and trailing space
                splitValues[0] = splitValues[0].strip()
            lastQuote = re.search(r"'\s*$", splitValues[-1]) # ... so here we use search instead
            if lastQuote:
                splitValues[-1] = splitValues[-1][:lastQuote.start()]
            # XXX given that "''" is not a valid separator, should it be considered as escaping "'"
            # and thus replaced in the split tokens?
            return splitValues
        else:
            # String should be in the format "1 2 3". Here, direct usage of str.split(sep=None)
            # takes care of ignoring multiple separating spaces.
            return string.split()

    @staticmethod
    def getParamTypeName(paramType):
        return paramType.__name
    
    @staticmethod
    def valueOf(typeName):
        return ParamType.__values[typeName]

def parseSingleBool(x):
    """
    Parse a single value of type BOOLEAN, according to the E2E-ICD rules.
    """
    x = x.strip()
    if x == "TRUE":
        return True
    elif x == "FALSE":
        return False
    else:
        raise ValueError('Accepted values are TRUE or FALSE (case-sensitive), got "{0}"'.format(x))

def parseSingleFile(fileName):
    """
    Parse the value of a single item in a FILE or FOLDER parameter. If the value is an absolute path
    under the current platform, it is returned as-is, whereas a relative path is composed with the
    "base folder" determined from environment variables per the E2E-ICD.
    See filesBaseDir.
    """
    # Check whether the fileName has absolute path or relative
    f = pathlib.PurePath(fileName)
    if f.is_absolute():
        return str(f) # Also performs / -> \ slash conversion in Windows
    base = filesBaseDir()
    if not f.parts and not base.parts:
        return '' # Empty base and fileName: return '' instead of '.', keeping previous behaviour
    return str(base / fileName)

def filesBaseDir(env=None):
    """
    Returns the "base directory" used to resolve relative paths of ParamType.FILE and
    ParamType.FOLDER types. Following the E2E-ICD, it is read from the enviroment variable
    "E2E_HOME".	As a deprecated option, if that variable is not present, the "OSFI_HOME" variable
    is consulted next. If both are absent from the environment, then the current working directory is used.
    
    :param env: The default (None) uses the default environment, but it can be replaced (e.g. for
        unit tests) by passing it in.
    :return: Base directory used to resolve relative paths in parameters.
    :rtype: pathlib.PurePath
    """
    if env is None:
        env=get_env()
    ENVKEY_BASE_DIR = "Parameter.base_dir"
    stored = env[ENVKEY_BASE_DIR]
    if stored is not None:
        return stored
    
    e2e_home = env.external_var("E2E_HOME")
    legacy_home = env.external_var("OSFI_HOME")
    if e2e_home is not None:
        newBdir = pathlib.PurePath(e2e_home)
    elif legacy_home is not None:
        Logger.warning("OSFI: E2E_HOME is not set, but OSFI_HOME is. Note that OSFI_HOME is deprecated and may be removed.")
        newBdir = pathlib.PurePath(legacy_home)
    else:
        Logger.warning("OSFI: E2E_HOME is not set, using the current directory for relative paths.")
        newBdir = pathlib.PurePath('')
    env[ENVKEY_BASE_DIR] = newBdir
    return newBdir

#
# Define the user-visible parameter types
#
ParamType.INTEGER = ParamType("INTEGER", parseFun=int, is_str=False)
ParamType.FLOAT   = ParamType("FLOAT", parseFun=float, is_str=False)
ParamType.BOOLEAN = ParamType("BOOLEAN", parseFun=parseSingleBool, is_str=False)
ParamType.STRING  = ParamType("STRING", parseFun=lambda x: x, is_str=True) # No parsing required for strings
ParamType.FILE    = ParamType("FILE", parseFun=parseSingleFile, is_str=True)
ParamType.FOLDER  = ParamType("FOLDER", parseFun=parseSingleFile, is_str=True)
ParamType.TIME    = ParamType("TIME", parseFun=TimeValue.parse, is_str=False)

File: sceps-codes/SCEPScodes/OSFI/ParamType.m
classdef ParamType
%
% openSF Integration Libraries (OSFI)
% Deimos Space, S.L.U.
% 
% This file is part of OSFI. OSFI is free software; you can redistribute it
% and/or modify it under the terms of the 'ESA Software Community Licence Permissive' as
% published by the European Space Agency; either version 2.4 of the License,
% or (at your option) any later version. You should have received a
% copy of the 'ESA Software Community Licence Permissive - v2.4' along with this program
% or one can be found at <http://eop-cfi.esa.int/index.php/docs-and-mission-data/licensing-documents>.
%
    enumeration
        INTEGER, FLOAT, BOOLEAN, STRING, FILE, FOLDER, TIME
    end

    methods
        function val = parseValues(obj, txt)
            % PARSEVALUES receives either a string or a cell array of strings
            % and returns the parsed value according to the parameter type.
            % If given a cell array, the output will be an array or cell, with
            % the same shape as the input.
            switch obj % Each branch must define val _and_ parseErrors (or return on its own)
            case {ParamType.INTEGER, ParamType.FLOAT}
                val = str2double(txt);
                parseErrors = isnan(val) & ~strcmpi(txt, 'nan');
                if obj == ParamType.INTEGER
                    % The range mandated in the E2E-ICD is that of Java int = Matlab int32. Type double
                    % has enough precision to represent all int32 values exactly, so if any fpVal ~= val
                    % (conversion doesn't roundtrip), it was not a valid input. This catches both 
                    % out-of-range integers and floating point values.
                    fpVal = val;
                    val = int32(val);
                    parseErrors = parseErrors | fpVal ~=val;
                end

            case ParamType.BOOLEAN % NOTE: per the E2E-ICD, only capitals!
                val = strcmp(txt, 'TRUE');
                % Check that the false values in val are actually FALSE
                shouldBeNotVal = strcmp(txt, 'FALSE');
                parseErrors = (val == shouldBeNotVal); % Fields that are neither TRUE nor FALSE
            
            case ParamType.TIME
                if iscellstr(txt)
                    % Parse into a cell, storing empty values in any unparseable inputs
                    val = cellfun(@(s) TimeValue.parse(s), txt, 'UniformOutput', false, ...
                            'ErrorHandler', @(id, msg, idx) []);
                    parseErrors = cellfun(@isempty, val);
                    if ~any(parseErrors(:)) % Recast into a TimeValue object array
                        val = reshape([val{:}], size(val));
                    end
                else
                    val = TimeValue.parse(txt);
                    return % Error checking performed by the function itself
                end

            case ParamType.STRING % No-op
                val = txt;
                return

            case {ParamType.FILE, ParamType.FOLDER}
                val = ParamType.getFileFormattedValues(txt);
                return % Error checking is performed by the function itself

            otherwise
                error('Unknown parameter type %s', char(obj));
            end

            %% Check for errors and report any with the offending values
            if any(parseErrors(:)) % Logical matrix with the same dimensions as txt/val
                error('Invalid value(s) for %s: %s', char(obj), ...
                        ParamType.parseErrorBadValues(txt, parseErrors));
            end
        end

        function split = splitValues(obj, txt, expect_empty)
            % SPLITVALUES splits a string depending on the parameter type. Numeric types
            % and BOOLEAN are split on whitespace, while the string-like types expect a
            % set of quoted values separated by whitespace. The output is always a cell
            % array of the tokens.
            txt = strtrim(txt);
            if isempty(txt) && nargin > 2 && expect_empty
                split = {}; % For empty nodes in ARRAY-typed parameters, since regex split would return {''}
            elseif obj.nonScalarsAreQuoted()
                % In this case we don't want to collapse delimiters, since '' is a valid element
                split = strsplit(txt, '''\s+''', 'CollapseDelimiters', false, 'DelimiterType', 'RegularExpression');
                if ~isempty(split)
                    % Since we have split on "' '", we need to remove the begin-quote from
                    % the first item and the end-quote from the last one.
                    split{1}   = regexprep(split{1},   '^''', '');
                    split{end} = regexprep(split{end}, '''$', '');
                end
            else
                % Split on whitespace, collapse multiple delimiters
                split = strsplit(txt);
            end
        end

        function isquoted = nonScalarsAreQuoted(obj)
            switch obj
            case {ParamType.INTEGER, ParamType.FLOAT, ParamType.BOOLEAN, ParamType.TIME}
                isquoted = false;

            case {ParamType.STRING, ParamType.FILE, ParamType.FOLDER}
                isquoted = true;
            otherwise
                error('Unknown parameter type %s', char(obj));
            end
        end
    end

    methods (Static=true)
        function str = parseErrorBadValues(txt, badMask)
            if ~iscellstr(txt) % Single value
                str = ['"' txt '"'];
                return
            end
            % Non scalar: build a string with (indices)="value" with all the invalid values
            badVals = txt(badMask);
            badLinIdx = find(badMask);
            dims = size(badMask);
            if isvector(badMask) % Avoid matrix indices for vectors
                dims = length(badMask); 
            end
            formattedBadVals = cell(1,numel(badVals));
            for iBad = 1:numel(badVals)
                subs = cell(1,length(dims));
                [subs{:}] = ind2sub(dims, badLinIdx(iBad));
                formattedBadVals{iBad} = sprintf('(%s)="%s"', ...
                        strjoin(cellfun(@num2str, subs, 'UniformOutput', false), ','), badVals{iBad});
            end
            str = [ 'at ' strjoin(formattedBadVals, ', ')];
        end

        function ffv = getFileFormattedValues(val)
            ffv = val;
            if ispc % Windows accepts both separators so normalize to \
                ffv = strrep(ffv, '/', '\');
            end
            
            % If ffv is a string, this function converts from string to char in the input (to then call cellfun)
            inputStr = isstring(val);
            if inputStr
                val = convertStringsToChars(val);
            end

            % Check if any of the path(s) is not absolute (no such function in Matlab)
            is_abs_fun = @(f) logical(javaMethod('isAbsolute', javaObject('java.io.File', f)));
            if iscell(val)
                isabs = cellfun(is_abs_fun, val);
            else
                isabs = is_abs_fun(val);
            end
            if all(isabs)
                return % XXX we might want to perform separator conversion here?
            end
            % If any relative path, apply the base dir from environment variables, to them (only!)
            base = ParamType.getFilesBaseDir(OsfiEnv.getInstance);
            if ~isempty(base)
                base_path = javaMethod('toPath', javaObject('java.io.File', base));
                make_abs = @(f) char(base_path.resolve(f));
                if iscell(val)
                    ffv(~isabs) = cellfun(make_abs, ffv(~isabs), 'UniformOutput', false);
                elseif inputStr
                    % If ffv is a string, convert from char to string
                    ffv = string(make_abs(val));
                else
                    ffv = make_abs(val);
                end
            end
        end
        
        function base = getFilesBaseDir(env)
            % GETFILESBASEDIR returns the "base directory" used to resolve
            % relative paths of FILE and FOLDER types. Per the the E2E-ICD,
            % it is read from the enviroment variable E2E_HOME. As a
            % deprecated option, if that variable is not present, the
            % OSFI_HOME variable is consulted next. If both are absent from
            % the environment, then the current working directory is used.
            ENVKEY_BASE_DIR = 'Parameter.base_dir';
            if nargin == 0
                env = OsfiEnv.getInstance;
            end
            storedVal = env(ENVKEY_BASE_DIR);
            if iscellstr(storedVal) && length(storedVal) == 1
                base = storedVal{1};
                return
            end
            
            e2e_home = env.getExternalVar('E2E_HOME');
            legacy_home = env.getExternalVar('OSFI_HOME');
            if ~isempty(e2e_home)
                base = e2e_home;
            elseif ~isempty(legacy_home)
                warning('OSFI:ConFM:Deprecated', ['E2E_HOME is not set, but OSFI_HOME is. '...
                        'Note that OSFI_HOME is deprecated and may be removed.']);
                base = legacy_home;
            else
                warning('OSFI:ConFM:HOME_UNDEF', ...
                        'E2E_HOME is not set, using the current folder for relative paths.');
                base = '';
            end
            env(ENVKEY_BASE_DIR) = {base}; % Prevent '' from being swallowed
        end
    end
end
